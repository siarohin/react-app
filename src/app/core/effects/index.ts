import { combineEpics } from "redux-observable";

import { movieEffects } from "./movies.effect";
import { userPreferencesEffects } from "./user-preferences.effect";
import { toastEffects } from "./toast.effect";

export * from "./common.util";

export const effects = combineEpics(movieEffects, userPreferencesEffects, toastEffects);
