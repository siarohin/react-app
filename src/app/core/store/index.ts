export * from "./reducers";
export { SharedModels } from "./shared";
export { MoviesActions, MoviesModels } from "./movies";
export { UserPreferencesActions, UserPreferencesModels } from "./user-preferences";
export { ToastActions } from "./toast";
