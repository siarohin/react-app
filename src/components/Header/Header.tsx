import React, { ReactElement } from "react";

export const Header = ({ title }: { title: string }): ReactElement => <h1 className="app-header">{title}</h1>;
