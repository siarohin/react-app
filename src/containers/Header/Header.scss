.app-header__top-right {
  grid-row: 1 / span 1;
  grid-column: 2 / span 1;
  justify-self: flex-end;
}

.app-header__container {
  grid-row: 2 / span 1;
  grid-column: 1 / span 2;
  justify-self: center;
  width: 100%;
  height: calc(3 * var(--app-spacing-xxxl));
}
