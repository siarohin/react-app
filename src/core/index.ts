export * from "./constants";
export * from "./services";
export * from "./effects";
export * from "./store";
export * from "./store-configuration";
