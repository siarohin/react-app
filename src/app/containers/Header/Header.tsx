import React, { ComponentType, ReactElement } from "react";
import { useHistory } from "react-router-dom";
import { connect } from "react-redux";

import "./Header.scss";
import { DialogAction, SharedModels, State, UserPreferencesActions, UserPreferencesModels } from "../../core";
import { IMovieUpsertAction } from "../../shared";
import { getPath } from "../../utils";
import { AddMovieButton, SearchPanel } from "../../components";

const Header = (
  props: UserPreferencesModels.ISearch & IMovieUpsertAction & SharedModels.IDispatchAction
): ReactElement => {
  const { selected, dispatch } = props;
  const history = useHistory();

  const handleClick = (): void => {
    const dialogAction: UserPreferencesModels.IMovieAction = {
      action: DialogAction.CREATE
    };
    dispatch(UserPreferencesActions.updateDialogAction({ dialogAction }));
  };

  const handleSubmit = (searchValue: string): void => {
    const value: string = searchValue?.trim();

    dispatch(UserPreferencesActions.updateSearchValue({ selected: value }));
    history.push(getPath(value));
  };

  return (
    <>
      <div className="app-header__top-right">
        <AddMovieButton handleClick={handleClick} />
      </div>
      <div className="app-header__container">
        <SearchPanel selected={selected} onSubmit={handleSubmit} />
      </div>
    </>
  );
};

const mapStateToProps = (state: State): UserPreferencesModels.ISearch => ({
  selected: state.userPreferences.search.selected
});

export default connect(mapStateToProps)(Header as ComponentType<any>);
