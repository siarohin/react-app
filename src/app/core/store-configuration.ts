import { applyMiddleware, createStore, compose, Store } from "redux";
import { combineEpics, createEpicMiddleware, Epic } from "redux-observable";

import { reducers, State } from "./store";
import { isServer } from "./constants";

function getComposeEnhancers(): any {
  if (!isServer) {
    return (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
  }
  return compose;
}

export interface AppStore {
  store: Store;
  runEffects: (effects$?: Array<Epic>) => void;
}

/**
 * Store configuration
 */
export function configureStore(initialState: State = {} as State): AppStore {
  const epicMiddleware = createEpicMiddleware();
  const composeEnhancers = getComposeEnhancers();
  const middlewares = [epicMiddleware];

  const store: Store = createStore(reducers, initialState, composeEnhancers(applyMiddleware(...middlewares)));

  return {
    store,
    runEffects: (effects$?: Array<Epic>) =>
      effects$
        ? epicMiddleware.run(combineEpics(...effects$))
        : // eslint-disable-next-line no-console
          console.warn("Effects are not running, no epics array was provided")
  };
}
