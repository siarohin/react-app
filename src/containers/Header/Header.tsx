import React, { ComponentType, ReactElement } from "react";
import { connect } from "react-redux";

import "./Header.scss";
import { DialogAction, SharedModels, State, UserPreferencesActions, UserPreferencesModels } from "../../core";
import { AddMovieButton, SearchPanel } from "../../components";
import { IMovieUpsertAction } from "../../models";

const Header = (
  props: UserPreferencesModels.ISearch & IMovieUpsertAction & SharedModels.IDispatchAction
): ReactElement => {
  const { selected, editableAction, dispatch } = props;

  const handleClick = (): void => editableAction({ action: DialogAction.CREATE });

  const handleSubmit = (value: string): void => {
    dispatch(UserPreferencesActions.updateSearchValue({ selected: value }));
  };

  return (
    <>
      <div className="app-header__top-right">
        <AddMovieButton handleClick={handleClick} />
      </div>
      <div className="app-header__container">
        <SearchPanel selected={selected} onSubmit={handleSubmit} />
      </div>
    </>
  );
};

const mapStateToProps = (state: State): UserPreferencesModels.ISearch => ({
  selected: state.userPreferences.search.selected
});

export default connect(mapStateToProps)(Header as ComponentType<any>);
