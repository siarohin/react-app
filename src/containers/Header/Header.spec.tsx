import React from "react";
import { Provider } from "react-redux";
import renderer, { ReactTestRenderer } from "react-test-renderer";
import configureStore from "redux-mock-store";

import Header from "./Header";
import { State } from "../../core";

const mockStore = configureStore([]);

describe("Containers.Header.Header: ", () => {
  let store: any;
  let component: ReactTestRenderer;

  beforeEach(() => {
    store = mockStore({
      userPreferences: {
        search: {
          selected: ""
        }
      }
    } as State);

    component = renderer.create(
      <Provider store={store}>
        <Header />
      </Provider>
    );
  });

  it("should render component", () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
});
