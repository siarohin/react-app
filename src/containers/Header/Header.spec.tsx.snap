// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Containers.Header.Header:  should render component 1`] = `
Array [
  <div
    className="app-header__top-right"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-text app-add-movie-button"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
         Add movie
      </span>
    </button>
  </div>,
  <div
    className="app-header__container"
  >
    <div
      className="app-search-panel"
    >
      <h3
        className="app-search-panel__name"
      >
        Find your movie
      </h3>
      <form
        className="app-search-panel__form"
        onSubmit={[Function]}
      >
        <div
          className="MuiFormControl-root MuiTextField-root app-search-panel__form-field"
        >
          <div
            className="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiFilledInput-input"
              disabled={false}
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="What do you want to watch?"
              required={false}
              type="text"
              value=""
            />
          </div>
        </div>
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-text app-search-panel__form-button"
          disabled={false}
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="submit"
        >
          <span
            className="MuiButton-label"
          >
            Search
          </span>
        </button>
      </form>
    </div>
  </div>,
]
`;
